File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left, right = -1, len(nums)
        first = -1

        while left + 1 < right:
            mid = left + (right - left) // 2

            if nums[mid] >= target:
                right = mid
            else:
                left = mid
        
        first = right
        
        left, right = -1, len(nums)

        while left + 1 < right:
            mid = left + (right - left) // 2

            if nums[mid] <= target:
                left = mid
            else:
                right = mid

        second = left

        if (first or second) == -1 or second < first:
            return [-1, -1]
        return [first, second]
            

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int left = 0;
        int right = nums.length-1;
        int start, end = -1;

        while (left <= right){
            int mid = left + (right - left) / 2;

            if (nums[mid] >= target){
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        start = left;

        left = 0;
        right = nums.length-1;

        while (left <= right){
            int mid = left + (right - left) / 2;

            if (nums[mid] <= target){
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        end = right;

        if (start > end || start == -1 || end == -1){
            return new int[] {-1, -1};
        }

        return new int[] {start, end};
    }
}